axonserver/src/main/java/io/axoniq/axonserver/message/event/EventDispatcher.java

```diff
@@ -250,44 +254,66 @@ public void listAggregateEvents(String context, Authentication principal, GetAgg
-                        .tag("context", context)
-                        .tag("stream", "aggregate_events")
-                        .tag("origin", "client_request")
-                        .metrics();
+                                                     AtomicLong lastSeq = contextView.get(LAST_SEQ_KEY);
+                                                     boolean allowedSnapshot =
+                                                             request.getAllowSnapshots()
+                                                                     && lastSeq.get() == -1;
+
+                                                     GetAggregateEventsRequest newRequest = request
+                                                             .toBuilder()
+                                                             .setAllowSnapshots(
+                                                                     allowedSnapshot)
+                                                             .setInitialSequence(lastSeq.get() + 1)
+                                                             .build();
+
+                                                     logger.debug("Reading events from seq#{} for aggregate {}",
+                                                                  lastSeq.get() + 1,
+                                                                  request.getAggregateId());
+                                                     return eventStore.aggregateEvents(context, principal, newRequest);
+                                                 }
+                                )
+                                .limitRate(aggregateEventsPrefetch * 5, aggregateEventsPrefetch)
+                                .doOnEach(signal -> {
+                                    if (signal.hasValue()) {
+                                        ((AtomicLong) signal.getContextView().get(LAST_SEQ_KEY))
+                                                .set(signal.get().getAggregateSequenceNumber());
+                                    }
+                                })
+                                .retryWhen(retrySpec
+                                                   .doBeforeRetry(t -> logger.warn(
+                                                           "Retrying to read events aggregate stream due to {}:{}, for aggregate: {}",
+                                                           t.failure().getClass().getName(),
+                                                           t.failure().getMessage(),
+                                                           request.getAggregateId())))
+                                .timeout(Duration.ofMillis(aggregateEventsTimeoutMilliseconds))
+                                .doOnNext(m -> logger.trace("event {} for aggregate {}", m, request.getAggregateId()))
+                                .doOnEach(signal -> {
+                                    if (signal.hasError()) {
```

Comment:
I would move this code to doOnError instead checking for error in doOnEach signal